common/logging/src/tracing_logging_layer.rs

```diff
@@ -0,0 +1,137 @@
+use std::collections::HashMap;
+use std::fs::{self, create_dir_all, File, OpenOptions};
+use std::io::Write;
+use std::path::PathBuf;
+use std::sync::mpsc::{self, Sender};
+use std::thread;
+
+lazy_static! {
+    pub static ref TRACING_LOGGING_DEPENDENCIES: Vec<String> =
+        vec!["libp2p".to_string(), "discv5".to_string()];
+}
+
+/// Layer that handles `INFO`, `WARN` and `ERROR` logs emitted per dependency and
+/// writes them to a file. Dependencies are enabled via the `RUST_LOG` env flag.
+pub struct LoggingLayer {
+    pub file_writer_streams: HashMap<String, NonBlockingFileWriter>,
+}
+
+impl<S: tracing_core::Subscriber> tracing_subscriber::layer::Layer<S> for LoggingLayer {
+    fn on_event(
+        &self,
+        event: &tracing_core::Event<'_>,
+        _ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) {
+        let meta = event.metadata();
+
+        let target = match meta.target().split_once("::") {
+            Some((crate_name, _)) => crate_name,
+            None => "unknown",
+        };
+
+        let Some(file_writer) = self.file_writer_streams.get(target) else {
+            return;
+        };
+
+        let mut visitor = LogMessageExtractor {
+            message: String::default(),
+        };
+
+        event.record(&mut visitor);
+
+        match *meta.level() {
+            tracing_core::Level::INFO => {
+                let _ = file_writer.write(visitor.message);
+            }
+            tracing_core::Level::WARN => {
+                let _ = file_writer.write(visitor.message);
+            }
+            tracing_core::Level::ERROR => {
+                let _ = file_writer.write(visitor.message);
+            }
+            _ => {}
+        }
```

Comment:
why are you filtering out debug messages?

Reply:
no real reason, I'll add debug messages

Reply:
don't think you should be filtering here by level, since that can be controlled with the `RUST_LOG` flag